Alex and Happy Part 1/article.md

```diff
@@ -0,0 +1,817 @@
+─────►│ 0 ├────►││1│├────►││2││     │
+      │   │     │└─┘│     │└─┘│     │
+      └───┘     └─┬─┘     └───┘     │
+                  │                 │
+                  │f                │
+                  ▼                 │
+                ┌───┐               │
+                │┌─┐│ [a-z]         │
+                ││3│├───────────────┘
+                │└─┘│
+                └───┘
+```
+
+To understand how to use the diagram above, imagine that this is a city where each numbered square is a place where you can drive to, and each arrow represents a one-way street.
+The arrow coming from nowhere to the square with 0 is where you start driving, and squares with another square inside are places where you can park your car. In this case, you are not allowed to park it in the square with 0.
+
+The arrows also have directions in them, guided by your GPS. Your GPS may tell you to drive straight into an `i`, then turn at an `[a-z]`, and then to stop, for example. However, your GPS may also ask you to take illegal turns, for example, to turn at an `@`, which will always lead you to a dead end, where you will be stuck.
+
+To refine the terminology, the city is the _automaton_ (plural: _automata_) represented by a _state transition graph_, each numbered square is called a _state_, and each arrow is called a _transition_.
+When a transition comes from nowhere and into a state, we say that this state is a _start state_, which is state 0 in this example.
+The places were you can park are called _accepting states_, i.e., 1, 2, 3, and 4.
+
+The GPS represents the _input_ (string) of the automaton.
+The dead ends represent an implicit _error state_ present in every DFA, which you can think of like a hidden state number 5, with all invalid transitions from every state to it.
+
+To simplify matters, on ambiguous transitions such as `i` and `[a-z]`, assume that the one with a single letter takes precedence.
+
+For example, to check the expression "int", we will follow this path: 0 → 1 → 2 → 4.
+
+Now, we need to attribute meanings to each state so that this automaton can be useful, and we will do so like this:
+* 0: Error.
+* 1: Identifier.
+* 2: In.
+* 3: If.
+* 4: Identifier.
+* 5 (implicit): Error.
+
+Since we stopped at 4, the result is that "int" is an identifier, so we get a token that is `Identifier "int"`.
+The resulting token would be `In` if we stopped at 2 instead (for the expression "in").
+
+What if we consumed no input and stopped at state 0, or the input was something like "Foo123" where we'd reach the implicit state 5?
+In such cases, we halt and indicate a lexical error.
+
+As stated previously, we won't write our lexical analyzer from scratch. Instead, we will use Alex to generate a lexer that creates a stream of tokens, which the Happy-generated parser will parse.
+
+### Regular expressions
+
+Alex uses regular expressions (regexes) to define patterns, we'll keep it pretty simple and explain some of them which we'll use.
```

Comment:
Is this regular regex? I have doubts whether it is necessary to explain it beyond a link.

Reply:
Yes and no. Alex uses mostly standard notation, but it has a few quirks.

Reply:
I'd prefer to keep this section. I don't know how it's usually done, but in my university, I remember we learned regex during our compiler classes for the lexing part, so I think it makes sense to explain it to the user here.

Reply:
Sure, that's ok.